flow/include/flow/Traceable.h

```diff
@@ -0,0 +1,247 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FLOW_TRACEABLE_H
+#define FLOW_TRACEABLE_H
+#pragma once
+
+#include <stdarg.h>
+#include <stdint.h>
+#include <string>
+#include <string_view>
+#include <type_traits>
+
+#define PRINTABLE_COMPRESS_NULLS 0
+
+template <class IntType>
+char base16Char(IntType c) {
+	switch ((c % 16 + 16) % 16) {
+	case 0:
+		return '0';
+	case 1:
+		return '1';
+	case 2:
+		return '2';
+	case 3:
+		return '3';
+	case 4:
+		return '4';
+	case 5:
+		return '5';
+	case 6:
+		return '6';
+	case 7:
+		return '7';
+	case 8:
+		return '8';
+	case 9:
+		return '9';
+	case 10:
+		return 'a';
+	case 11:
+		return 'b';
+	case 12:
+		return 'c';
+	case 13:
+		return 'd';
+	case 14:
+		return 'e';
+	case 15:
+		return 'f';
+	default:
+		std::abort();
+	}
+}
+
+// forward declare format from flow.h as we
+// can't include flow.h here
+std::string format(const char* form, ...);
+
+template <class T, class T2 = void>
+struct Traceable : std::false_type {};
+
+#define FORMAT_TRACEABLE(type, fmt)                                                                                    \
+	template <>                                                                                                        \
+	struct Traceable<type> : std::true_type {                                                                          \
+		static std::string toString(type value) { return format(fmt, value); }                                         \
+	}
+
+FORMAT_TRACEABLE(bool, "%d");
+FORMAT_TRACEABLE(signed char, "%d");
+FORMAT_TRACEABLE(unsigned char, "%d");
+FORMAT_TRACEABLE(short, "%d");
+FORMAT_TRACEABLE(unsigned short, "%d");
+FORMAT_TRACEABLE(int, "%d");
+FORMAT_TRACEABLE(unsigned, "%u");
+FORMAT_TRACEABLE(long int, "%ld");
+FORMAT_TRACEABLE(unsigned long int, "%lu");
+FORMAT_TRACEABLE(long long int, "%lld");
+FORMAT_TRACEABLE(unsigned long long int, "%llu");
+FORMAT_TRACEABLE(float, "%g");
+FORMAT_TRACEABLE(double, "%g");
+FORMAT_TRACEABLE(void*, "%p");
+FORMAT_TRACEABLE(volatile long, "%ld");
+FORMAT_TRACEABLE(volatile unsigned long, "%lu");
+FORMAT_TRACEABLE(volatile long long, "%lld");
+FORMAT_TRACEABLE(volatile unsigned long long, "%llu");
+FORMAT_TRACEABLE(volatile double, "%g");
+
+template <class Enum>
+struct Traceable<Enum, std::enable_if_t<std::is_enum_v<Enum>>> : std::true_type {
+	static std::string toString(Enum e) { return format("%lld", (int64_t)e); }
+};
```

Comment:
I added this to make it possible to convert enum class to strings using their integer value.